### DISCUSSION TITLE
How to enter negative speed with ESF-MEAS

### DISCUSSION THREAD

**Discussion by FSet89** [semuconsulting/pyubx2#164] [Q&A] [answered]
**Describe the bug**
- Device Model/Generation: u-blox F9R
- pyubx2 version: 1.2.22

I can't send a negative speed using the ESF-MEAS message:
```
import pyubx2
pyubx2.UBXMessage("ESF", "ESF-MEAS", pyubx2.SET, timeTag=1000, numMeas=1, dataField_01=-10, dataType_01=11)
```

Traceback (most recent call last):
  File "/usr/lib/python3.7/site-packages/pyubx2/ubxmessage.py", line 107, in _do_attributes
    offset, pdict, key, index, **kwargs
  File "/usr/lib/python3.7/site-packages/pyubx2/ubxmessage.py", line 165, in _set_attribute
    att, offset, index, **kwargs
  File "/usr/lib/python3.7/site-packages/pyubx2/ubxmessage.py", line 219, in _set_attribute_group
    offset, attd, key1, index, **kwargs
  File "/usr/lib/python3.7/site-packages/pyubx2/ubxmessage.py", line 157, in _set_attribute
    att, offset, index, **kwargs
  File "/usr/lib/python3.7/site-packages/pyubx2/ubxmessage.py", line 333, in _set_attribute_bitfield
    self._payload += bitfield.to_bytes(bfs, "little")
OverflowError: can't convert negative int to unsigned

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.7/site-packages/pyubx2/ubxmessage.py", line 82, in __init__
    self._do_attributes(**kwargs)
  File "/usr/lib/python3.7/site-packages/pyubx2/ubxmessage.py", line 131, in _do_attributes
    ) from err
pyubx2.exceptions.UBXTypeError: Overflow error for attribute 'group' in SET message class ESF-MEAS

**Comment by semuadmin** [comment]
Hi @FSet89 

According to the [F9R Integration Manual](https://content.u-blox.com/sites/default/files/ZED-F9R_Integrationmanual_UBX-20039643.pdf), the ESF-MEAS `dataField` attribute is an **unsigned** 24-bit integer (U24), so this is expected behaviour.

<img width="756" alt="Screenshot 2024-10-31 at 16 00 29" src="https://github.com/user-attachments/assets/ca57cdd3-bacb-4b06-b114-98d49c529c06">

I recommend having another look at the [F9R Integration Manual](https://content.u-blox.com/sites/default/files/ZED-F9R_Integrationmanual_UBX-20039643.pdf) for clarification of how to use the `dataField` and `dataType` attributes.

**Comment by FSet89** [comment]
Hi, you're right, however according to the application notes the speed is signed (section 6.1.2): https://content.u-blox.com/sites/default/files/documents/ZED-F9R-GettingStarted_AppNote_UBX-22035176.pdf

**Comment by semuadmin** [comment, marked as answer]
> Hi, you're right, however according to the application notes the speed is signed (section 6.1.2): https://content.u-blox.com/sites/default/files/documents/ZED-F9R-GettingStarted_AppNote_UBX-22035176.pdf

OK so clearly a discrepancy between different u-blox documentation sources!

In point of fact, different ESF-MEAS `dataTypes` have different attribute types (some are signed, some are unsigned, some use individual bitflags), so the `dataField` attribute is actually implemented in pyubx2 as generic X24 (24-bit binary data). This means that, to enter a -ve number, you have to use what's known as the ["two's complement"](https://en.wikipedia.org/wiki/Two%27s_complement) binary representation of the number, e.g. -10 in 24-bit two's complement  format is `0b111111111111111111110110`, derived as follows:

1. start with the +ve number: 10 in 24-bit binary = `0b000000000000000000001010`
2. invert all the bits (1 becomes 0, 0 becomes 1), which gives you `0b111111111111111111110101`
3. add 1, which gives you `0b111111111111111111110110`

In Python terms, this is equivalent to `-10 & 0b111111111111111111111111`.

So the following amended code snippet should work:

```python
from pyubx2 import UBXMessage, SET

value = -10
value_as_twoscomp = value & 0b111111111111111111111111
print(f"{value_as_twoscomp:024b}")

msg = UBXMessage(
    "ESF",
    "ESF-MEAS",
    SET,
    timeTag=1000,
    numMeas=1,
    dataField_01=value_as_twoscomp,
    dataType_01=11,
)

print(msg)

```
```
111111111111111111110110
<UBX(ESF-MEAS, timeTag=1000, timeMarkSent=0, timeMarkEdge=0, calibTtagValid=0, numMeas=1, id=0, dataField_01=16777206, dataType_01=11)>
```

FYI https://portal.u-blox.com/s/question/0D5Oj00000XUqA9KAL/ubx-esfmeas-how-to-enter-ve-speed-value-in-datafield

**Reply by semuadmin** [reply]
@FSet89  FYI https://github.com/semuconsulting/pyubx2/discussions/168
